dvc/stage/cache.py

```diff
@@ -210,51 +212,46 @@ def restore(self, stage, run_cache=True, pull=False):
         )
         cached_stage.checkout()
 
-    @staticmethod
-    def _transfer(func, from_remote, to_remote):
-        ret = []
+    def transfer(self, from_odb, to_odb):
+        from dvc.fs._callback import FsspecCallback
+
+        from_fs = from_odb.fs
+        to_fs = from_odb.fs
+        func = _log_exceptions(fs.utils.copy)
+        runs = from_fs.path.join(from_odb.fs_path, "runs")
 
-        runs = from_remote.fs.path.join(from_remote.fs_path, "runs")
-        if not from_remote.fs.exists(runs):
-            return []
+        ret = []
+        if not from_fs.exists(runs):
+            return ret
 
-        from_path = from_remote.fs.path
-        for src in from_remote.fs.find(runs):
-            rel = from_path.relpath(src, from_remote.fs_path)
-            dst = to_remote.fs.path.join(to_remote.fs_path, rel)
-            key = to_remote.fs.path.parent(dst)
+        for src in from_fs.find(runs):
+            rel = from_fs.path.relpath(src, from_odb.fs_path)
+            dst = to_fs.path.join(to_odb.fs_path, rel)
+            key = to_fs.path.parent(dst)
             # check if any build cache already exists for this key
             # TODO: check if MaxKeys=1 or something like that applies
             # or otherwise this will take a lot of time!
-            if to_remote.fs.exists(key) and first(to_remote.fs.find(key)):
+            if to_fs.exists(key) and first(to_fs.find(key)):
                 continue
-            func(src, dst)
-            ret.append(
-                (from_path.name(from_path.parent(src)), from_path.name(src))
-            )
 
+            src_name = from_fs.path.name(src)
+            parent_name = from_fs.path.name(from_fs.path.parent(src))
+            with FsspecCallback.as_tqdm_callback(
+                desc=src_name,
+                bytes=True,
+                total=-1,
+            ) as cb:
+                func(from_fs, src, to_fs, dst, callback=cb)
```

Comment:
This now creates a Tqdm progress bar, and uses `fs.utils.copy` instead of `fs.upload`/`fs.download`.